source/validate_instruction.cpp

```diff
@@ -27,474 +27,72 @@
-    EXECUTION_MODE_CASE2(Triangles, Geometry, Tessellation);
-    EXECUTION_MODE_CASE(InputTrianglesAdjacency, Geometry);
-    EXECUTION_MODE_CASE(Quads, Tessellation);
-    EXECUTION_MODE_CASE(Isolines, Tessellation);
-    EXECUTION_MODE_CASE2(OutputVertices, Geometry, Tessellation);
-    EXECUTION_MODE_CASE(OutputPoints, Geometry);
-    EXECUTION_MODE_CASE(OutputLineStrip, Geometry);
-    EXECUTION_MODE_CASE(OutputTriangleStrip, Geometry);
-    EXECUTION_MODE_CASE(VecTypeHint, Kernel);
-    EXECUTION_MODE_CASE(ContractionOff, Kernel);
-    case SpvExecutionModeLocalSize:
-      break;
-  }
-#undef EXECUTION_MODE_CASE
-#undef EXECUTION_MODE_CASE2
-  return SPV_SUCCESS;
-}
-
-#define DIM_CASE(DIM, CAPABILITY)                                   \
-  case SpvDim##DIM:                                                 \
-    if (_.hasCapability(SpvCapability##CAPABILITY) == false) {      \
-      return _.diag(SPV_ERROR_INVALID_CAPABILITY)                   \
-             << "Dim " #DIM " requires " #CAPABILITY " capability"; \
-    }                                                               \
-    break
+}  // namespace anonymous
 
-spv_result_t DimCapabilityCheck(ValidationState_t& _, SpvDim dim) {
-  switch (dim) {
-    DIM_CASE(1D, Sampled1D);
-    DIM_CASE(Cube, Shader);
-    DIM_CASE(Rect, SampledRect);
-    DIM_CASE(Buffer, SampledBuffer);
-    DIM_CASE(SubpassData, InputAttachment);
-    case SpvDim2D:
-    case SpvDim3D:
-      break;
-  }
-#undef DIM_CASE
-  return SPV_SUCCESS;
-}
+namespace libspirv {
 
-spv_result_t SamplerAddressingModeCapabilityCheck(
-    ValidationState_t& _, SpvSamplerAddressingMode sampler_addressing_mode) {
-  std::string mode;
-  switch (sampler_addressing_mode) {
-    case SpvSamplerAddressingModeNone:
-      mode = "None";
-      break;
-    case SpvSamplerAddressingModeClampToEdge:
-      mode = "ClampToEdge";
-      break;
-    case SpvSamplerAddressingModeClamp:
-      mode = "Clamp";
-      break;
-    case SpvSamplerAddressingModeRepeat:
-      mode = "Repeat";
-      break;
-    case SpvSamplerAddressingModeRepeatMirrored:
-      mode = "RepeatMirrored";
-      break;
-  }
-  if (_.hasCapability(SpvCapabilityKernel) == false) {
+spv_result_t CapCheck(ValidationState_t& _,
+                      const spv_parsed_instruction_t* inst) {
+  spv_opcode_desc opcode_desc;
+  if (SPV_SUCCESS == _.grammar().lookupOpcode(inst->opcode, &opcode_desc) &&
+      !_.HasAnyOf(opcode_desc->capabilities))
     return _.diag(SPV_ERROR_INVALID_CAPABILITY)
-           << mode + " sample address mode requires Kernel capability";
+           << "Opcode " << spvOpcodeString(inst->opcode)
+           << " requires one of these capabilities: "
+           << ToString(opcode_desc->capabilities, _.grammar());
+  for (int i = 0; i < inst->num_operands; ++i) {
+    spv_operand_desc operand_desc;
+    if (SPV_SUCCESS ==
+            _.grammar().lookupOperand(inst->operands[i].type,
```

Comment:
This is correct for most operands.  It is incorrect when the operand is the bitwise-OR of masks.

See "Set 5" in the definition of spv_operand_type_t.

For example, SPV_OPERAND_TYPE_IMAGE.  The Bias bit requires Shader, but Lod depends on nothing, and MinLod depends on MinLod.